### PR TITLE
[Datasets] [Docs] Update `map_batches` documentation

### PR DESCRIPTION
**Pull request by bveeramani**
Signed-off-by: Balaji Veeramani <balaji@anyscale.com>

**Depends on**:
- [x] #28434 

## Why are these changes needed?

* Users aren't aware of `BatchMapper`
* Users can't figure out how to write UDFs, and there's no link to the relevant user guide
* Users get confused by the examples
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/dataset.py
Comment (clarkzinzow) [inline python/ray/data/dataset.py:323]:
```suggestion
        Batches are represented as dataframes, ndarrays, or lists. The default batch
```

File: python/ray/data/dataset.py
Comment (clarkzinzow) [inline python/ray/data/dataset.py:336]: Might want to qualify this with "If using Ray AIR for training and/or batch inference,"

File: python/ray/data/dataset.py
Comment (clarkzinzow) [inline python/ray/data/dataset.py:344]: Quite the pair! 😄

File: python/ray/data/dataset.py
Comment (clarkzinzow) [inline python/ray/data/dataset.py:361]: Might want to do something more interesting than a cast for the example UDF?

**Review by clarkzinzow — COMMENTED** [review]
Nice, big improvement!

File: python/ray/data/dataset.py
Comment (bveeramani) [reply python/ray/data/dataset.py:361]: Do you have any suggestions? I tried thinking of something that's both simple and concise (like, few lines of code for both the dataframe and `map_fn`).

Comment (clarkzinzow) [reply python/ray/data/dataset.py:361]: Hmm ideally it would be something like generating a new column via a simple linear combination of one or more other columns, e.g.
```python
def map_fn(batch: pd.DataFrame) -> pd.DataFrame:
    batch["age_in_dog_years"] = 7 * batch["age"]
    return batch
```

Comment (xwjiang2010) [reply python/ray/data/dataset.py:361]: I also have ~2 users unsure about whether the change is gonna be on the same column or different.

They were staring at `ds.map_batches(lambda batch: [v * 2 for v in batch])`.

Comment (bveeramani) [reply python/ray/data/dataset.py:361]: @clarkzinzow updated example with your recommendation.

> I also have ~2 users unsure about whether the change is gonna be on the same column or different.

@xwjiang2010 could you elaborate? I'm not sure if I understand.

Comment (xwjiang2010) [reply python/ray/data/dataset.py]: I guess they just didn't realize that it's up to them (the way how they write the udf) to decide whether the new value is applied to the old column or new one. 

So in this case, they can write it in a way that "age" becomes 7 times the original value "in place" or just add a new column with the new age value. 

@clarkzinzow btw, I do that maths all the time with my cats as well!! :)

File: python/ray/data/dataset.py
Comment (clarkzinzow) [inline python/ray/data/dataset.py:361]: Supernit: Self-linking in the docstring isn't necessary, since they're already here.
```suggestion
            of data. ``ds.map_batches`` applies the function in parallel.
```

File: python/ray/data/dataset.py
Comment (clarkzinzow) [inline python/ray/data/dataset.py:355]: Nit: I think that a simple `.default_batch_format` reference should work, and is a bit more concise/readable in the non-rendered docs. `.default_batch_format` would include the `.` in the displayed text, while `~.default_batch_format` would remove it; I think that the former is better since it makes it more obvious that it's a method on the current class. Lmk what you think.
```suggestion
            Call :meth:`.default_batch_format` to determine the default batch type.
```

File: python/ray/data/dataset.py
Comment (clarkzinzow) [inline python/ray/data/dataset.py:387]: Nit: More concise.
```suggestion
            :class:`~ray.data.ActorPoolStrategy`.
```

File: python/ray/data/dataset.py
Comment (clarkzinzow) [inline python/ray/data/dataset.py:403]: Nit: I think that our lint would add a comma here.
```suggestion
            ...     num_gpus=1,
```

**Review by clarkzinzow — APPROVED** [review]
This is a massive improvement to this docstring! Only small nits left, so this looks good to merge after the default batch PR is merged and this is rebased.

File: python/ray/data/dataset.py
Comment (bveeramani) [reply python/ray/data/dataset.py:355]: When I rendered the documentation, the `.` wasn't included.

image

In any case, `.default_batch_format` is definitely more concise, so I've applied the suggestion. Didn't know you could do that

Comment (clarkzinzow) [reply python/ray/data/dataset.py]: Hmm interesting, the `.` must be implicitly sliced off. Good to know!

File: python/ray/data/dataset.py
Comment (shawnpanda) [inline python/ray/data/dataset.py:374]: Could you also add an example where the data is arrow blocks? Thanks!

Comment (bveeramani) [reply python/ray/data/dataset.py:374]: Do you mean an example where the underlying block type is Arrow, or the input batch format is Arrow?

Comment (shawnpanda) [reply python/ray/data/dataset.py:374]: The input batch format is arrow

Comment (bveeramani) [reply python/ray/data/dataset.py:374]: I don't think you can get an Arrow input batch? AFAIK they always get converted to pandas.

```
def to_default(self) -> Block:
        if self.is_tensor_wrapper():
            default = self.to_numpy()
        else:
            # Always promote Arrow blocks to pandas for consistency, since
            # we lazily convert pandas->Arrow internally for efficiency.
            default = self.to_pandas()
        return default
```

Comment (bveeramani) [reply python/ray/data/dataset.py:374]: Oh, you mean like when you manually specify `batch_format=arrow`?

Comment (shawnpanda) [reply python/ray/data/dataset.py:374]: Hi, Balaji, yes, you are right in that arrows get converted into pandas under the hood. However, as a user, I initially didn't know that and was looking for an example for the arrow input format. Maybe you can add a sentence somewhere to indicate that users with arrow input format can just use whatever panda uses since it's converted under the hood? 

The context is that I was using pymongoarrow [here](https://mongo-arrow.readthedocs.io/en/latest/quickstart.html) and was under the impression that the returned format is in arrow. Therefore, I wanted to use dataset to format it.

Tagging @jjyao  and @jianoaix  in case my description is unclear.

Comment (bveeramani) [reply python/ray/data/dataset.py:374]: > The context is that I was using pymongoarrow [here](https://mongo-arrow.readthedocs.io/en/latest/quickstart.html) and was under the impression that the returned format is in arrow. Therefore, I wanted to use dataset to format it.

Would it be correct to say that you were expecting the batch format to be an Arrow table because the underlying block type is Arrow?

My understanding is that the underlying block type is more of an implementation detail. Consumers of `read_*` APIs don't need to know the underlying block type; they just need to know just the batch format (related: https://github.com/ray-project/ray/pull/28434).

If the block type is an implementation detail, a sentence like "If the underlying block type is Arrow..." might be unhelpful or confusing to the majority of users.

@clarkzinzow what do you think?

Comment (clarkzinzow) [reply python/ray/data/dataset.py:374]: Hmm we already call out that the default batch format for tabular data (both Arrow and Pandas blocks) is the pandas DataFrame batch format in the docs for the `batch_format` parameter, we delve further into this in the "Transforming Datasets" feature guide, and we know have the `ds.default_batch_format()` API so users can query this at runtime, but maybe we need an explicit example or call out here for the Arrow --> Pandas case that makes this extra clear. But something akin to "Arrow blocks will yield pandas DataFrame batches by default" is also dipping into implementation details, as @bveeramani said, so not sure if we want to do that.

We have a few options here:
1. Have both a `ray.data.from_pandas()` example and a `ray.data.from_arrow()` example, showing that each yield a `pd.DataFrame` batch format by default.
2. Have a single example that's not tied to pandas or Arrow, e.g. `ray.data.read_parquet()` or `ray.data.read_csv()`, emphasizing that the abstract concept of "tabular" data is what leads to the default `pd.DataFrame` batch, which encompasses pandas DataFrames and Arrow Tables
3. Have a `ray.data.from_pandas()` example but have a sentence that a Dataset created from Arrow data will yield a `pd.DataFrame` batch by default (doesn't have to be a full example).
4. (2), and also have a sentence explicitly calling out that a Dataset created from Arrow data will yield a `pd.DataFrame` batch by default.

@shawnpanda Would all of these had made it clear that the default batch format for Arrow tabular data is still `pd.DataFrame`?

Comment (shawnpanda) [reply python/ray/data/dataset.py:374]: Hi Balaji and Clark, thanks for the responses here

> Would all of these have made it clear that the default batch format for Arrow tabular data is still pd.DataFrame?

I have a preference for 1, 3, 4 so that the examples are more relatable to what I was working on and might be directly copy-able. The context is that I am very new to datasets and actually don't understand the nuances between different data formats. You guys can decide on a concise option and will defer to you guys. Don't want to make this great doc improvement PR feel like dragging on.

File: python/ray/data/dataset.py
Comment (jianoaix) [inline python/ray/data/dataset.py:340]: This is not necessarily, when it's lazy it doesn't do actual work.

File: python/ray/data/dataset.py
Comment (jianoaix) [inline python/ray/data/dataset.py:454]: nit: to *learn*? User cannot choose default type.

File: python/ray/data/dataset.py
Comment (jianoaix) [inline python/ray/data/dataset.py:341]: I'm sure how useful a time complexity is here; the algorithmic aspect is not interesting for map_batches (and most other APIs). And the `p` is not even supplied in function args and is going to invite user questions.

Comment (bveeramani) [reply python/ray/data/dataset.py:341]: Okay, I can remove it. It was in the original docstring so I kept it.

File: python/ray/data/dataset.py
Comment (bveeramani) [reply python/ray/data/dataset.py:454]: I think "determine" is okay in this context.

This is one of the Google definitions:
> ascertain or establish exactly, typically as a result of research or calculation.
> "the point of our study was to determine what is true, not what is practicable"

File: python/ray/data/dataset.py
Comment (bveeramani) [reply python/ray/data/dataset.py:340]: I can remove this, too. It was in the original docstring so I assumed it to be true.

Comment (bveeramani) [reply python/ray/data/dataset.py:340]: Addressed

File: python/ray/data/dataset.py
Comment (bveeramani) [reply python/ray/data/dataset.py:341]: Address

**Review by jianoaix — APPROVED** [review]

File: python/ray/data/dataset.py
Comment (bveeramani) [reply python/ray/data/dataset.py:326]: Fixed

**Pull request merged**
